homeassistant/components/livisi/binary_sensor.py

```diff
@@ -0,0 +1,110 @@
+"""Code to handle a Livisi Binary Sensor."""
+from __future__ import annotations
+
+from typing import Any
+
+from homeassistant.components.binary_sensor import (
+    BinarySensorDeviceClass,
+    BinarySensorEntity,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN, LIVISI_STATE_CHANGE, LOGGER, WDS_DEVICE_TYPE
+from .coordinator import LivisiDataUpdateCoordinator
+from .entity import LivisiEntity
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up binary_sensor device."""
+    coordinator: LivisiDataUpdateCoordinator = hass.data[DOMAIN][config_entry.entry_id]
+    known_devices = set()
+
+    @callback
+    def handle_coordinator_update() -> None:
+        """Add Window Sensor."""
+        shc_devices: list[dict[str, Any]] = coordinator.data
+        entities: list[BinarySensorEntity] = []
+        for device in shc_devices:
+            if device["id"] not in known_devices and device["type"] == WDS_DEVICE_TYPE:
+                livisi_binary: BinarySensorEntity = LivisiWindowDoorSensor(
+                    config_entry, coordinator, device
+                )
+                LOGGER.debug("Include device type: %s", device["type"])
+                coordinator.devices.add(device["id"])
+                known_devices.add(device["id"])
+                entities.append(livisi_binary)
+        async_add_entities(entities)
+
+    config_entry.async_on_unload(
+        coordinator.async_add_listener(handle_coordinator_update)
+    )
+
+
+class LivisiBinarySensor(LivisiEntity, BinarySensorEntity):
+    """Represents a Livisi Binary Sensor."""
+
+    def __init__(
+        self,
+        config_entry: ConfigEntry,
+        coordinator: LivisiDataUpdateCoordinator,
+        device: dict[str, Any],
+        capabilityName: str,
+    ) -> None:
+        """Initialize the Livisi sensor."""
+        super().__init__(config_entry, coordinator, device)
+        self._capability_id = self.capabilities[capabilityName]
+
+    async def async_added_to_hass(self) -> None:
+        """Register callbacks."""
+        await super().async_added_to_hass()
+
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass,
+                f"{LIVISI_STATE_CHANGE}_{self._capability_id}",
+                self.update_states,
+            )
+        )
+
+    @callback
+    def update_states(self, state: bool) -> None:
+        """Update the state of the device."""
+        self._attr_is_on = state
+        self.async_write_ha_state()
+
+
+class LivisiWindowDoorSensor(LivisiBinarySensor):
+    """Represents a Livisi Window/Door Sensor as a Binary Sensor Entity."""
```

Comment:
I'm not sure you need to have two classes (`LivisiWindowDoorSensor` and `LivisiBinarySensor`)
If you have only "window/door" sensors, then it's not useful.

If you are planning to have more binary sensor types, then maybe it is better to add two right away and maybe look at `EntityDescription`.

Reply:
Yes, there are / will be other binary sensors.
At least one of the "battery low" will require a completely different implementation and be part of multiple devices. As you like to split features and code quality changes/refactoring into separate PRs, I thought it might be smart to plan for this in advance.

Reply:
OK - hopefully this will work for the other binary sensors.